openmicroscopy/openmicroscopy
# Cors settings

# What this PR does

--cors

This replaces #5252 and instead of installing ```django-cors-headers``` app, it merely adds the essential settings, allowing users to install https://github.com/ottoyiu/django-cors-headers and configure remaining settings themselves.

The relies on #5269 to allow setting of middleware.

# Test setup:

These steps should be performed and deployed by devspace at http://10.0.51.146/web/api/
so you can skip this for testing.

1. Install app: ``` pip install django-cors-headers```
2. Add app: ``` bin/omero config append omero.web.apps '"corsheaders"' ```
    - NB: This is called for in app docs above, but doesn't seem to be necessary?!
3. Add middleware. Middleware order with respect to existing middleware must follow [rules](https://github.com/ottoyiu/django-cors-headers), e.g:
```
bin/omero config append omero.web.middleware '{"index": 0.5, "class": "corsheaders.middleware.CorsMiddleware"}'
bin/omero config append omero.web.middleware '{"index": 10, "class": "corsheaders.middleware.CorsPostCsrfMiddleware"}'
```

4. Configure middleware to allow connections from specified origins:
```
bin/omero config append omero.web.cors_origin_whitelist '"localhost:8000"'
```
OR from all origins
```
bin/omero config set omero.web.cors_origin_allow_all True
```

# Test CORS connection:

 - Checkout this repo to get the example code:

```
cd openmicroscopy/examples/Training/javascript/

python -m SimpleHTTPServer
```

 - Go to http://localhost:8000/, open devtools and follow the Network tab to see what is happening.
 - Enter the url http://10.0.51.146/web/ and click "Connect" (this devspace deploy has been configured to accept "localhost:8000" as shown above)
 - This will GET available servers, CSRF token and login url and show a login screen.
 - Login uses POST (should see the pre-flight OPTIONS request) and then lists Projects.
 - Connecting from origins other than "localhost:8000" should fail. e.g. different port:
``` python -m SimpleHTTPServer 8080```



# Related reading

Link to cards, tickets, other PRs:

1. background for understanding this PR

2. what this PR assists, fixes, or otherwise affects
